File: src/apiClient.js
const API_BASE = import.meta.env.VITE_API_BASE_URL || "/api";

async function request(path, options) {
  const response = await fetch(`${API_BASE}${path}`, options);
  if (!response.ok) {
    let message = `Request failed: ${response.status}`;
    try {
      const data = await response.json();
      if (data?.error) {
        message = data.error;
      }
    } catch {
      // no-op
    }
    throw new Error(message);
  }
  return response.json();
}

export async function fetchContent() {
  return request("/content");
}

export async function fetchAdminContent() {
  return request("/admin/content");
}

export async function updateAdminContent(payload) {
  return request("/admin/content", {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload)
  });
}

export async function createOrder(payload) {
  return request("/orders", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload)
  });
}

export async function createContactMessage(payload) {
  return request("/contact-messages", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload)
  });
}

File: flutter_app/lib/services/api_client.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../models/content_models.dart';

class ApiClient {
  ApiClient(this.baseUrl);

  final String baseUrl;

  Future<SiteContent> fetchContent() async {
    final json = await _request('/content');
    return SiteContent.fromJson(json);
  }

  Future<AdminContent> fetchAdminContent() async {
    final json = await _request('/admin/content');
    return AdminContent.fromJson(json);
  }

  Future<void> updateAdminContent(Map<String, dynamic> payload) async {
    await _request('/admin/content', method: 'PUT', body: payload);
  }

  Future<void> createOrder(Map<String, dynamic> payload) async {
    await _request('/orders', method: 'POST', body: payload);
  }

  Future<void> createContactMessage(Map<String, dynamic> payload) async {
    await _request('/contact-messages', method: 'POST', body: payload);
  }

  Future<Map<String, dynamic>> _request(
    String path, {
    String method = 'GET',
    Map<String, dynamic>? body,
  }) async {
    final uri = Uri.parse('$baseUrl$path');
    late http.Response response;

    switch (method) {
      case 'POST':
        response = await http.post(
          uri,
          headers: {'Content-Type': 'application/json'},
          body: jsonEncode(body),
        );
        break;
      case 'PUT':
        response = await http.put(
          uri,
          headers: {'Content-Type': 'application/json'},
          body: jsonEncode(body),
        );
        break;
      default:
        response = await http.get(uri);
    }

    final payload = response.body.isEmpty
        ? <String, dynamic>{}
        : jsonDecode(response.body);

    if (response.statusCode < 200 || response.statusCode > 299) {
      final message =
          payload is Map<String, dynamic> && payload['error'] != null
          ? '${payload['error']}'
          : 'Request failed: ${response.statusCode}';
      throw Exception(message);
    }

    return payload is Map<String, dynamic>
        ? payload
        : <String, dynamic>{'data': payload};
  }
}
